### PR TITLE
EntityExtinguishEvent

### LINKED ISSUE #5143: EntityExtinguishEvent
Currently there is no way to prevent an entity's fire ticks being reset to zero when the entity enters water or is in the rain.  It would be helpful if there was an EntityExtinguishEvent that is fired when an entity enters water or rain while on fire.  It would be possible to cancel, and you would be able to get the entity that is being extinguished and also the cause of the entity being extinguished.  As far as I know, there is currently no way to accomplish this.  I have tried adding to the fire ticks when an entity (in this case a player) is in water, but it is just being reset to zero instantly.  For 1.17, this event would also need to be fired when an entity enters powder snow.

### PR DESCRIPTION
**Pull request by MrIvanPlays**
Closes #5143 

~~There are some calls which would return a reason of UNKNOWN because I was not sure what reason it was. 
And also I'm unsure for some of the reasons' names. Please check these 2~~

### CODE REVIEW

File: Spigot-Server-Patches/0673-EntityExtinguishEvent.patch [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
++import io.papermc.paper.event.entity.EntityExtinguishEvent; // Paper
+ import java.util.Arrays;
+ import java.util.Collection;
+ import java.util.Collections;
+@@ -557,7 +558,22 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
+     }
+ 
+     public void extinguish() {
++        // Paper start - call EntityExtinguishEvent
++        extinguish(EntityExtinguishEvent.Reason.UNKNOWN);
```
Comment (Proximyst) [inline Spigot-Server-Patches/0673-EntityExtinguishEvent.patch]: Prefer something like PLUGIN here

File: Spigot-Server-Patches/0673-EntityExtinguishEvent.patch [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
++    }
++    public void extinguish(EntityExtinguishEvent.Reason reason) {
++        if (EntityExtinguishEvent.getHandlerList().getRegisteredListeners().length != 0) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0673-EntityExtinguishEvent.patch]: This is not an event that requires such a check

File: Spigot-API-Patches/0274-EntityExtinguishEvent.patch [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
++
++    public enum Reason {
++        /**
++         * The entity has died.
++         */
++        DEATH,
++        /**
++         * The entity has collided with a block of cauldron filled with water.
++         */
++        COLLISION,
```
Comment (Proximyst) [inline Spigot-API-Patches/0274-EntityExtinguishEvent.patch]: Does this also affect splash potions of water, water bottles, and the lingering pots, if any of these at all extinguish you?

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0274-EntityExtinguishEvent.patch (hunk 3 above)
Comment (MrIvanPlays) [reply Spigot-API-Patches/0274-EntityExtinguishEvent.patch]: this is one of the names im really unsure of as it is called when extinguish() is getting called in BlockCauldron#a(IBlockData, World, BlockPosition, Entity)

the event is only called when extinguish() is being called, but I also might track setFireTicks(0) to call it there too if you want.

File: Spigot-Server-Patches/0673-EntityExtinguishEvent.patch (hunk 1 above)
Comment (MrIvanPlays) [reply Spigot-Server-Patches/0673-EntityExtinguishEvent.patch]: I changed it to PLUGIN but further looking at it , extinguish() doesn't have it's own method in the API so it doesn't really make sense to be PLUGIN

File: Spigot-API-Patches/0274-EntityExtinguishEvent.patch (hunk 3 above)
Comment (MrIvanPlays) [reply Spigot-API-Patches/0274-EntityExtinguishEvent.patch]: Looks like splash potions, bottles and lingering potions of water don't call extinguish() at all, neither setFireTicks(0) so they probably don't extinguish you. 
either way we need a new name for this...

**Comment by MrIvanPlays** [comment]
I've currently not set any reason of 3 extinguish() calls ( the reason of them is PLUGIN ) 

line counts are with accuracy of the code after these patches are applied:
Entity#entityBaseTick() line 441 
EntityHuman#attack(Entity) line 1211
EntityLiving#entityBaseTick() line 288

anyone having any suggestions for a reason here?

**Comment by kennytv** [comment]
> I've currently not set any reason of 3 extinguish() calls ( the reason of them is PLUGIN )
> 
> line counts are with accuracy of the code after these patches are applied:
> Entity#entityBaseTick() line 441
> EntityHuman#attack(Entity) line 1211
> EntityLiving#entityBaseTick() line 288
> 
> anyone having any suggestions for a reason here?

The one in the entity base tick is basically just the fire extinguishing naturally (= BURN_OUT, albeit faster due to fire immunity, still just from counting down). 
For the base tick in EntityLiving, I'd suggest adding an enum entry like `FIRE_IMMUNITY`.
Not sure about the attack, since it's instantly putting out the just set fire from a fire aspect attack that has been cancelled - meaning there has not *really* been a fire in the first place.

Also in general on the `COLLISION` reason; that name sounds unfitting imo, since you're not being put out by *colliding with a cauldron*, but by going into the cauldron with water, so something like `IN_WATER_CAULDRON` would be both less confusing and more descriptive.

File: Spigot-API-Patches/0274-EntityExtinguishEvent.patch [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
++    public boolean isCancelled() {
++        return cancelled;
++    }
++
++    /**
++     * {@inheritDoc}
++     */
++    @Override
++    public void setCancelled(boolean cancel) {
++        this.cancelled = true;
```
Comment (kashike) [inline Spigot-API-Patches/0274-EntityExtinguishEvent.patch]: No. Use the parameter.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0274-EntityExtinguishEvent.patch (hunk 4 above)
Comment (MrIvanPlays) [reply Spigot-API-Patches/0274-EntityExtinguishEvent.patch]: oops ill fix l8r

**Comment by ObbyEquipped** [comment]
I compiled this and tried it out.  It seems like there is an issue where you will still not take fire damage while in water even if the extinguish event is cancelled.  I have looked through the code, but I can't figure out why.

**Comment by MrIvanPlays** [comment]
> I compiled this and tried it out. It seems like there is an issue where you will still not take fire damage while in water even if the extinguish event is cancelled. I have looked through the code, but I can't figure out why.

did u verify that the event is in fact getting called? 
there's also the theory that this may be clientsided /shrug

**Comment by ObbyEquipped** [comment]
> > I compiled this and tried it out. It seems like there is an issue where you will still not take fire damage while in water even if the extinguish event is cancelled. I have looked through the code, but I can't figure out why.
> 
> did u verify that the event is in fact getting called?
> there's also the theory that this may be clientsided /shrug

I verified that the event is actually being called.  I think that this could very well be a client-side issue.  However, that would be a problem for Mojang if it is possible to avoid fire damage by convincing your client that you are in water.

**Comment by MrIvanPlays** [comment]
i mean we can convince the client it is not in water or in whatever makes the fire go off but idk if any of paper's maintainers would want such a mechanic...

**Comment by ObbyEquipped** [comment]
I think that we should examine the code more to see what could be causing this.  Maybe there is a simple solution.

**Comment by aurorasmiles** [comment]
In EntityHuman there is another extinguish that isn't handled here, around line 1210

**Comment by Proximyst** [comment]
See #5242

**Pull request closed without merging**